server/src/main/java/org/apache/druid/server/coordinator/ClientCompactionRunnerInfo.java

```diff
@@ -0,0 +1,198 @@
+      CompactionEngine defaultCompactionEngine
+  )
+  {
+    CompactionEngine compactionEngine = newConfig.getEngine() == null ? defaultCompactionEngine : newConfig.getEngine();
+    if (compactionEngine == CompactionEngine.NATIVE) {
+      return new ValidationResult(true, null);
+    } else {
+      return msqEngineSupportsCompactionConfig(newConfig);
+    }
+  }
+
+  /**
+   * Checks if the provided compaction config is supported by MSQ. The following configs aren't supported:
+   * <ul>
+   * <li>finalizeAggregations set to false in context.</li>
+   *
+   * <li>partitionsSpec of type HashedParititionsSpec.</li>
+   *
+   * <li>maxTotalRows in DynamicPartitionsSpec.</li>
+   *
+   * <li>rollup set to false in granularitySpec when metricsSpec is specified.</li>
+   * </ul>
+   *
+   * @param newConfig The updated compaction config
+   * @return ValidationResult. The reason string is null if isValid() is True.
+   */
+  private static ValidationResult msqEngineSupportsCompactionConfig(DataSourceCompactionConfig newConfig)
+  {
+    if (newConfig.getTuningConfig() != null) {
+      ValidationResult partitionSpecValidationResult =
+          validatePartitionsSpec(newConfig.getTuningConfig().getPartitionsSpec());
+      if (!partitionSpecValidationResult.isValid()) {
+        return partitionSpecValidationResult;
+      }
+    }
+    if (newConfig.getGranularitySpec() != null) {
+      ValidationResult rollupValidationResult = validateRollup(
+          newConfig.getMetricsSpec(),
+          newConfig.getGranularitySpec().isRollup()
+      );
+      if (!rollupValidationResult.isValid()) {
+        return rollupValidationResult;
+      }
+    }
+    return new ValidationResult(true, null);
+  }
+
+  /**
+   * Validte that partitionSpec is either 'dynamic` or 'range', and if 'dynamic', ensure maxTotalRows is null.
+   */
+  public static ValidationResult validatePartitionsSpec(PartitionsSpec partitionsSpec)
+  {
+    if (!(partitionsSpec instanceof DimensionRangePartitionsSpec
+          || partitionsSpec instanceof DynamicPartitionsSpec)) {
+      return new ValidationResult(false, StringUtils.format(
+          "Invalid partition spec type[%s] for MSQ engine."
+          + " Type must be either DynamicPartitionsSpec or DynamicRangePartitionsSpec.",
+          partitionsSpec.getClass()
+      )
+      );
+    }
+    if (partitionsSpec instanceof DynamicPartitionsSpec
+        && ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows() != null) {
+      return new ValidationResult(false, StringUtils.format(
+          "maxTotalRows[%d] in DynamicPartitionsSpec not supported for MSQ engine.",
+          ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows()
+      ));
+    }
+    return new ValidationResult(true, null);
+  }
+
+  /**
+   * Validate rollup is set to false in granularitySpec when metricsSpec is specified.
+   */
+  public static ValidationResult validateRollup(AggregatorFactory[] metricsSpec, boolean isRollup) {
```

Comment:
These methods should not be public. Tests should only use `validateCompactionConfig` method.

Reply:
They were kept public to enable reuse in `MSQCompactionRunner#validateCompactionTask` checks.